cmd/portforward.go

```diff
@@ -0,0 +1,560 @@
+	wsshell "github.com/mendersoftware/go-lib-micro/ws/shell"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"golang.org/x/sys/unix"
+
+	"github.com/mendersoftware/mender-cli/client/deviceconnect"
+)
+
+const (
+	argBindHost    = "bind"
+	readBuffLength = 64 * 1024
+	localhost      = "127.0.0.1"
+)
+
+const (
+	ProtoTypePortForward       = 3
+	MessageTypePortForwardNew  = "new"
+	MessageTypePortForwardStop = "stop"
+	MessageTypePortForward     = "forward"
+)
+
+const (
+	propertyConnectionID = "connection_id"
+	propertyRemoteHost   = "remote_host"
+	propertyRemotePort   = "remote_port"
+	propertyProtocol     = "protocol"
+)
+
+var portForwardCmd = &cobra.Command{
+	Use:   "port-forward <DEVICE_ID> <LOCAL_PORT:REMOTE_PORT> ...",
+	Short: "Forward LOCAL_PORT on the current host to the REMOTE_PORT of the device",
+	Long: "This command supports both TCP and UDP port-forwarding.\n\n" +
+		"The port specification can be prefixed with \"tcp/\" or \"udp/\".\n" +
+		"If no prefix is specified, TCP is the default.\n\n" +
+		"REMOTE_PORT can also be specified in the form REMOTE_HOST:REMOT_PORT, making\n" +
+		"it possible to port-forward to third hosts running in the device's network.\n" +
+		"In this case, the specification will be LOCAL_PORT:REMOTE_HOST:REMOTE_PORT.\n\n" +
+		"You can specify the port mapping multiple times.",
+	Example: "  mender-cli port-forward DEVICE_ID 8000:8000 8010:8010\n" +
+		"  mender-cli port-forward DEVICE_ID udp/8000:8000\n" +
+		"  mender-cli port-forward DEVICE_ID tcp/8000:192.168.1.1:8000",
+	Args: cobra.MinimumNArgs(2),
+	Run: func(c *cobra.Command, args []string) {
+		cmd, err := NewPortForwardCmd(c, args)
+		CheckErr(err)
+		CheckErr(cmd.Run())
+	},
+}
+
+func init() {
+	portForwardCmd.Flags().StringP(argBindHost, "", localhost, "binding host")
+}
+
+const (
+	protocolTCP = "tcp"
+	protocolUDP = "udp"
+)
+
+type portMapping struct {
+	Protocol   string
+	LocalPort  uint
+	RemoteHost string
+	RemotePort uint
+}
+
+// PortForwardCmd handles the port-forward command
+type PortForwardCmd struct {
+	server       string
+	skipVerify   bool
+	deviceID     string
+	bindingHost  string
+	portMappings []portMapping
+	connections  map[string]chan *ws.ProtoMsg
+	running      bool
+	stop         chan struct{}
+	err          error
+}
+
+func getPortMappings(args []string) ([]portMapping, error) {
+	var err error
+	portMappings := []portMapping{}
+	for _, arg := range args {
+		remoteHost := localhost
+		protocol := protocolTCP
+		if strings.Contains(arg, "/") {
+			parts := strings.SplitN(arg, "/", 2)
+			if parts[0] == protocolTCP {
+				protocol = protocolTCP
+			} else if parts[0] == protocolUDP {
+				protocol = protocolUDP
+			} else {
+				return nil, errors.New("unknown protocol: " + parts[0])
+			}
+			arg = parts[1]
+		}
+		var localPort, remotePort int
+		if strings.Contains(arg, ":") {
+			parts := strings.SplitN(arg, ":", 3)
+			if len(parts) == 3 {
+				remoteHost = parts[0]
+				parts = parts[1:]
+			}
```

Comment:
+1 for inclusion of host in the port specification!

However, this is the opposite order of the one used in ssh, e.g. `remoteserver.com:8080:80` vs `8080:remoteserver.com:80`. Is it better to swap them so that we are command line compatible?

SSH also has `10.0.0.10:8080:80`, but it means "bind to 10.0.0.10", so quite different meaning.

Reply:
fixed, it is now `LOCAL_PORT:REMOTE_HOST:REMOTE_PORT`, with a simplified version `LOCAL_PORT:REMOTE_PORT`